h2o-algos/src/main/java/hex/glm/DispersionTask.java

```diff
@@ -0,0 +1,625 @@
+            _oneMinusAlpha = 1 - _alpha;
+            _oneOverPhiPower = 1.0 / Math.pow(_dispersionParameter, _oneMinusAlpha);
+            _oneOverPhiSquare = 1.0 / (_dispersionParameter * _dispersionParameter);
+            _oneOverPhi3 = _oneOverPhiSquare / _dispersionParameter;
+            _debugOn = parms._debugTDispersionOnly;
+            _oneOverDispersion = 1 / _dispersionParameter;
+            _alphaMinus1TLogDispersion = (_alpha - 1) * Math.log(_dispersionParameter);
+            _alphaTimesPI = _alpha * Math.PI;
+            _indexBound = parms._max_series_index;
+            _logDispersionEpsilon = Math.log(parms._tweedie_epsilon);
+            _computationAccuracy = new boolean[_nWVs];
+            _alphaMinus1OverPhi = (_alpha - 1) / _dispersionParameter;
+            _alphaMinus1SquareOverPhiSquare = _alphaMinus1OverPhi * _alphaMinus1OverPhi;
+            _calAll = calALL;
+        }
+
+        public static void setInfoIndices(Map<InfoColNames, Integer> infoColName2Ind, int constOffset, boolean weightPresent) {
+            int offset = weightPresent ? 3 : 2;
+            InfoColNames[] infoC = InfoColNames.values();
+            offset += constOffset;
+            int infoColLen = infoC.length;
+            for (int index = 0; index < infoColLen; index++) {
+                infoColName2Ind.put(infoC[index], index + offset);
+            }
+        }
+
+        public static void setConstIndices(Map<ConstColNames, Integer> constColName2Ind, int offset) {
+            ConstColNames[] constVal =
+                    ConstColNames.values();
+            int constantColNum = constVal.length;
+            for (int index = 0; index < constantColNum; index++)
+                constColName2Ind.put(constVal[index], index + offset);
+        }
+
+        public void map(Chunk[] chks) {
+            int chkLen = chks[0].len();
+            int jKIndMax = 0, jKL = 0, jKU = 0, djKL = 0, djKU = 0, d2jKL = 0, d2jKU = 0;
+            double wvMax = 0, dwvMax = 0, d2wvMax = 0, logZ = 0, sumWVj = 0, sumDWVj = 0, sumD2WVj = 0, oneOverSumWVj = 0;
+            _logLL = 0;
+            _dLogLL = 0;
+            _d2LogLL = 0;
+            _nobsLL = 0;
+            _nobsDLL = 0;
+            _nobsD2LL = 0;
+            _nobNegSum = 0;
+            double tempLL = 0, tempDLL = 0, tempD2LL = 0;
+            Map<ConstColNames, Integer> constColName2Ind = new HashMap<>();
+            Map<InfoColNames, Integer> infoColName2Ind = new HashMap<>();
+            setConstIndices(constColName2Ind, _weightPresent ? 3 : 2);
+            setInfoIndices(infoColName2Ind, constColName2Ind.size(), _weightPresent);
+            double weight;
+            for (int rInd = 0; rInd < chkLen; rInd++) {
+                double response = chks[0].atd(rInd);
+                weight = _weightPresent ? chks[WEIGHTIND].atd(rInd) : 1;
+                if (response >= 0) { // this part is only valid for response >= 0
```

Comment:
For variance power 1 < p < 2, response can be = 0 or > 0.
For variance power p > 2, response must be > 0.

Reply:
Oh good, so this condition should really be unnecessary since negative response should not be allowed this far in the GLM evaluation. 

We check it here:
https://github.com/h2oai/h2o-3/blob/726a05f00ee30edf9e4827a573ab83952793ce64/h2o-algos/src/main/java/hex/glm/GLM.java#L1014-L1021

but I think there should be one check here as well (just to make sure that the `_response.min() >= 0`):
https://github.com/h2oai/h2o-3/blob/726a05f00ee30edf9e4827a573ab83952793ce64/h2o-algos/src/main/java/hex/glm/GLM.java#L896-L897

in case we don't set the `fix_tweedie_variance_power`. And since we don't support using Tweedie instead of Gaussian (`p==0`) it should be safe to do that.

Reply:
This is tricky.  We only restrict tweedie variance power when user use maximum likelihood mode.  We restrict it to be > 1.  However during normal operation, we don't restrict variance power level.  It can be >= 0.  Hence, in this case, the response can be positive and negative.  Hence, we cannot add the restriction here.

Reply:
Thanks for clarification! Somehow I thought it has to be strictly greater than 1.